libyul/optimiser/LoadResolver.cpp

```diff
@@ -76,3 +105,62 @@ void LoadResolver::tryResolve(
 			if (inScope(*value))
 				_e = Identifier{locationOf(_e), *value};
 }
+
+void LoadResolver::tryEvaluateKeccak(
+	Expression& _e,
+	std::vector<Expression> const& _arguments
+)
+{
+	// Runtime saving: [PUSH32 32-byte-value] instead of [PUSH1 0 PUSH1 32 KECCAK256]
+	// Saves one KECCAK256 + one PUSH. Currently it is 30 + 6 + 3 = 39
+	unsigned runTimeGasSavings = GasCosts::keccak256Gas + GasCosts::keccak256WordGas + GasCosts::tier2Gas;
+	// Deploy time increase
+	// [PUSH32 32-byte-value]: 33 bytes
+	// instead of
+	// [PUSH1 0 PUSH1 32 KECCAK256]: 5 bytes
+	// Increase of 28 bytes.
+	// Current increase: 28 * 200: 56000
+	unsigned deployTimeGasLoss = 28 * GasCosts::createDataGas;
+	// We skip if the net savings are not positive
+	if (
+		m_expectedExecutionsPerDeployment.has_value() &&
```

Comment:
I didn't really think much about creation code. Why does it always pay off for creation? Even for `create` and `create2` inside contracts?

Reply:
Actually, I was talking nonsense. The opposite is true - for creation code you can always assume ``m_expectedExecutionsPerDeployment`` being 1 (since, well, the deploy code is executed once per deployment)... so the condition below would always be true, so if ``m_expectedExecutionsPerDeployment`` were supposed to mean creation code, then the condition would probably have to be ``!m_expectedExecutionsPerDeployment.has_value() || ...``....

Reply:
Why does it have to be `1`? Are you saying that `keccak` inside constructors should not be evaluated?

Reply:
Well, by definition, you execute the deploy code once per deployment, so the "expected executions per deployment" for deploy code is once.
I guess you're wondering about the case in which a contract keeps creating instances of another contract, so the deploy code is run several times in that sense? That's not really captured by "expectedExecutionsPerDeployment" I guess...

Reply:
By the way: you'd probably get around https://github.com/ethereum/solidity/pull/11026#discussion_r594225114 if you didn't evaluate during constructors.

Reply:
Independently, the gas costs would have to be calculated differently for deploy code.

Reply:
(e.g. in the inliner I added ``u256 GasMeter::dataGas(uint64_t _length, bool _inCreation, langutil::EVMVersion _evmVersion)`` to calculate the data gas cost of `_length` bytes depending on whether its creation code or not)